### PR TITLE
Multiple tokens

### LINKED ISSUE #18: Support multiple npm/optic tokens
We want to be able to allow multiple people to release npm packages. This means we need multiple NPM tokens and multiple Optic tokens (if needed).

- To do this, the action should support both of the above either as single tokens (as it does now) or multiple.
- Because action inputs cannot be arrays, we need to decide on an encoding format for arrays of values. Because those values are coming from repo secrets, it should also be able to store them there. I would suggest using comma separated values, which should be safe as neither npm or optic tokens should contain commas

Once we have a way to provide arrays of tokens, the action must:

- parse them correctly
- [zip](https://en.wikipedia.org/wiki/Zipping_(computer_science)) the two arrays and validate that they have the same cardinality
- when doing the npm publish, iterate through the pairs and implement the same behavior as [Promise.race](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/race)

### PR DESCRIPTION
**Pull request by sameer-coder**
This PR adds support for multiple npm/optic tokens.

**Usage:**
- You need to define npm and optic tokens in the Github secrets for each user that will receive the otp
- The `name` of the npm/optic secret for each user should start with `NPM_TOKEN_` and `OPTIC_TOKEN_`
- The `value` of the npm/optic secret for each user should be of the format `<github-username>:<token>`
- Default tokens can be provided with `NPM_TOKEN` and `OPTIC_TOKEN` secrets which will be used if individual user token secrets are not defined (backwards compatibility).

**How does it work?**
The action reads all secrets and checks if the value of a secret is starts with the `github.actor`. If it does then it uses that value for npm/optic token. If there are no matches and `NPM_TOKEN`/`OPTIC_TOKEN` secret is specified then this value is used instead.

Implementation is open for discussion.

Closes #18

### CODE REVIEW

File: README.md
Comment (simoneb) [inline README.md:44]: I'm not 100% sure that this is a viable option. This means that somebody using the action will have to trust sending all the secrets to the action. Before deciding that we don't have other options, have we considered alternatives?

Comment (sameer-coder) [reply README.md:44]: I suspected this might be an issue even though the action is public. I can try to filter the secrets using regex(there's an action for that) and only send the NPM/Optic token related secrets to the action.

Comment (simoneb) [reply README.md:44]: Yes, let's see if there's anything else here which could help https://docs.github.com/en/actions/learn-github-actions/expressions

File: README.md
Comment (simoneb) [inline README.md:56]: thanks for looking into ways to accomplish this. unfortunately, I don't think that asking users to include this code when using the action is a viable option, we need to come up with an alternative approach to support multiple tokens.

**Comment by Eomm** [comment]
I would propose an option.

As dependabot has its own `dependabod.yml`, we could define a `release-automation.yml` configuration file that we can load at startup. in that file, the user can list the users and the mapping between secrets' name and env var names so the dev is the director here too.

In this way, we access only what we need.

When maintainer-A run it, the action is still reading the maintainer-B tokens: I'm thinking if this could be optimized even more.

Example:

```
jobs:
  release:
    runs-on: ubuntu-latest
    steps:
      - uses: nearform/optic-release-automation-action@main
        with:
          config: './github/release-automation.yml'
          github-token: ${{secrets.github_token}}
          semver: ${{ github.event.inputs.semver }}
          npm-tag: ${{ github.event.inputs.tag }}
          env:
            KEY1: ${{ secrets.NPM_TOKEN_USER_FOO }}
            KEY2: ${{ secrets.OPTIC_TOKEN_FOO }}
            KEY_whatever: ${{ secrets.NPM_BAR }}
            KEY_hurra: ${{ secrets.OPTIC_TOKEN_BAR }}
```

And then the `release-automation.yml` file

```yml
users:
 - name: 'maintainer1'
   npm_token: 'KEY1'
   optic_token: 'KEY2'
 - name: 'maintainer-two'
   npm_token: 'KEY_whatever'
   optic_token: 'KEY_hurra'
```

**Comment by sameer-coder** [comment]
> I would propose an option.
> 
> As dependabot has its own `dependabod.yml`, we could define a `release-automation.yml` configuration file that we can load at startup. in that file, the user can list the users and the mapping between secrets' name and env var names so the dev is the director here too.
> 
> In this way, we access only what we need.
> 
> When maintainer-A run it, the action is still reading the maintainer-B tokens: I'm thinking if this could be optimized even more.
> 
> Example:
> 
> ```
> jobs:
>   release:
>     runs-on: ubuntu-latest
>     steps:
>       - uses: nearform/optic-release-automation-action@main
>         with:
>           config: './github/release-automation.yml'
>           github-token: ${{secrets.github_token}}
>           semver: ${{ github.event.inputs.semver }}
>           npm-tag: ${{ github.event.inputs.tag }}
>           env:
>             KEY1: ${{ secrets.NPM_TOKEN_USER_FOO }}
>             KEY2: ${{ secrets.OPTIC_TOKEN_FOO }}
>             KEY_whatever: ${{ secrets.NPM_BAR }}
>             KEY_hurra: ${{ secrets.OPTIC_TOKEN_BAR }}
> ```
> 
> And then the `release-automation.yml` file
> 
> ```yaml
> users:
>  - name: 'maintainer1'
>    npm_token: 'KEY1'
>    optic_token: 'KEY2'
>  - name: 'maintainer-two'
>    npm_token: 'KEY_whatever'
>    optic_token: 'KEY_hurra'
> ```

Sounds good to me. @simoneb what do you think?

**Comment by simoneb** [comment]
Thanks @Eomm @sameer-coder for the work and thinking on this. I think nonetheless that we're still not there, the proposals we have on the table so far all require considerable work in setting up this action from a user's perspective and we need to iterate on this proposals to come up with something better, which I am fairly confident we can do if we take a little more time to think on it.

**Comment by sameer-coder** [comment]
I can think of 2 other solutions which might make it easier for the end user.

- Enfore a naming convention for user npm/optic token. Example: `NPM_TOKEN_<github_user_id>`. Since GitHub user id is available in the action and is numeric it will be a valid secret name and should be easy to implement. Not sure if users will like the extra step of getting the user-id before saving the secret
- Maybe we can create a new action called `filter-secrets` which takes a GitHub username and secrets as input and returns the matching secret which is in our case is npm/optic token

@Eomm @simoneb Do you think any of these are feasible solutions?

**Comment by simoneb** [comment]
We need to do a bit of brainstorming on this. I'll schedule something

**Review by simoneb — COMMENTED** [review]
I like this new version! There's only one thing I would change. I would not introduce two new inputs, can we keep only npm-token and optic-token and delegate to the user to decide how to use them without imposing them to use a user-specific token and a generic one?

we could provide a few examples of usage:

- only default tokens, e.g. `npm-token: ${{secrets.NPM_TOKEN}}`
- only user-related tokens: e.g. `npm-token: ${{ secrets[format('NPM_TOKEN_{0}', github.actor)] }}` (note, no need to use env)
- both (please check this would work: `npm-token: ${{ secrets[format('NPM_TOKEN_{0}', github.actor)] ||  secrets.NPM_TOKEN }}`

File: action.js
Comment (Eomm) [inline action.js:12]: Are we still using the `user-optic-token`?

File: action.yml
Comment (Eomm) [inline action.yml:16]: This should be reverted

File: README.md
Comment (Eomm) [inline README.md:63]:
```suggestion
    *e.g.* npm-token: ${{ secrets.NPM_TOKEN }}
```

File: README.md
Comment (Eomm) [inline README.md:60]: What do you think moving this into the `example` section highlighting the "multiple user" feature.

I'm afraid here the user could miss it

Comment (sameer-coder) [reply README.md:60]: I think this is already in the example section. Are you suggesting we show separate(single and multi user) workflows in the example code?

Comment (sameer-coder) [reply README.md:60]: Or have a seperate multi-user section where we document this?

Comment (Eomm) [reply README.md:60]: Yes, this would give to this feature more attention I think

**Review by Eomm — APPROVED** [review]

File: README.md
Comment (simoneb) [inline README.md:59]: note that this is not mandatory, because if you don't specify those inputs simply there won't be any publish to npm, which is what we want because this action can be used to release other things (e.g. applications or github actions)

File: README.md
Comment (simoneb) [inline README.md:65]: let's be consistent with casing when mentioning npm and optic. let's use the same casing everywhere

File: README.md
Comment (simoneb) [inline README.md:69]: use `####` header and don't use "some examples" since the example is only one

File: README.md
Comment (simoneb) [inline README.md:92]: you can omit all this part since it's the same as the default example

**Review by simoneb — CHANGES_REQUESTED** [review]

**Review by simoneb — CHANGES_REQUESTED** [review]
There is no mention of the limitations of secret names for usernames which are not valid secret names

File: README.md
Comment (simoneb) [inline README.md:59]: Please always use correct casing! (github and otp which is an acronym)

```suggestion
- (Optional) You can define Npm and Optic tokens in GitHub secrets for each user that will receive the OTP. This is required only in case you want to publish to Npm.
```

File: README.md
Comment (simoneb) [inline README.md:65]: This is not entirely accurate. We don't enforce this naming convention in any way, it's just an option. Users could name their secrets whatever they want really! Here's we're providing an example of a way to use different tokens depending on the user who merged the pull request

File: README.md
Comment (simoneb) [reply README.md:70]: @sameer-coder please address this comment (simply accept the suggestion)

File: README.md
Comment (simoneb) [inline README.md:69]: the examples in this section must be formatted as code

**Review by simoneb — CHANGES_REQUESTED** [review]
See inline comments

File: README.md
Comment (sameer-coder) [reply README.md:68]: > @sameer-coder please address this comment (simply accept the suggestion)

I have formatted the examples as code. Is there anything else you feel should be changed here?

File: README.md
Comment (sameer-coder) [reply README.md]: You are right! The secret name can really be anything!
I have updated the docs

File: README.md
Comment (simoneb) [inline README.md:89]:
```suggestion
*Note*: Not all symbols that can be used in GitHub usernames are valid in secret names. One such example is the hyphen symbol (`-`). In such cases, this approach will not work.  
```

**Review by simoneb — APPROVED** [review]
I added one suggestion about the wording to use when mentioning the invalid symbols in secret names. Please include it before merging.

**Pull request merged**
